Auth beforefilter

ADmad opened the pull request:
Refs #6905

ADmad committed:
Remove uneeded implementedEvents().

ADmad committed:
Update docblock.

ADmad committed:
Allow running initial auth checks earlier.

If 'earlyAuth' config is set to true initial auth checks are done
in beforeFilter() allback instead of startup(), thus making user
info available in controller's beforeFilter().

ADmad committed:
Fix method visibility.

ADmad commented:
Suggestions welcome for a better config name.

dakota commented:
`legacyAuth` perhaps?

markstory commented on src/Controller/Component/AuthComponent.php:
```diff
@@ -147,7 +154,8 @@ class AuthComponent extends Component
         'unauthorizedRedirect' => true,
-        'storage' => 'Session'
+        'storage' => 'Session',
+        'earlyAuth' => false
```
Would it be crazy to have an option named `authenticateIn` that takes `beforeFilter` or `startup` as values?

ADmad replied:
Not crazy, I'll wait a bit before updating to allow others to weigh in.

jeremyharris replied:
Would it be possible to pass an event string that it registers? That way it could be almost anything.

ADmad replied:
@jeremyharris I don't think that would be easy as allowing any event would require using implementedEvents() and events get registered as soon as component instance is created before `Controller::initialize()` is called.

markstory commented on src/Controller/Component/AuthComponent.php:
```diff
@@ -281,18 +315,6 @@ public function startup(Event $event)
-        return [
-            'Controller.startup' => 'startup',
-        ];
-    }
```
Having this be explicit avoids the reflection calls (micro-optimization).

ADmad replied:
Will revert and add beforeFilter to the list.

ADmad committed:
Add back implementedEvents() to avoid reflection.

ADmad committed:
Allow triggering auth check on any configured event.

ADmad commented:
@dakota Thanks but I went with Mark's suggestion :)

@markstory Addressed your comments.

@jeremyharris The auth check can now be triggered on any configured event. Just for custom event you will have to take care of attaching AuthComponent as listener for that event yourself.

jeremyharris commented:
@ADmad oh cool, it could be done after all! Nice work!

markstory commented:
Looks good outside of PHPCS errors.

ADmad committed:
Fix CS error

ADmad commented:
CS errors fixed.
